### PR TITLE
Store all temporary files under a single folder

### LINKED ISSUE #2028: Store all files added to temporary folder below a single dedicated folder
### Describe the request

Arduino CLI stores the objects produced by compiling a sketch program under the temporary folder:

- `arduino-core-cache/` - Core objects
- `arduino-sketch-<hash>/` - Objects from sketch source files and library dependencies

These folders are intentionally left behind in the temporary folder after an Arduino CLI process exits so that they can be reused for subsequent compilations as applicable.

The proposal is for all Arduino tools to place any temporary files below a single dedicated folder. For example:

```text
tmp/
└── arduino/
    ├── arduino-core-cache/
    ├── arduino-sketch-33B7FB84F6F419C9D38C7A6D903E8E74/
    ├── arduino-sketch-8D027A8744130666283210C8D45F468C/
    ├──  arduino-sketch-BE89418D95A99EBE12879C9FE4080CB4/
    ...
```

🙂 User maintenance operations related to the files Arduino CLI adds to the temporary folder will be facilitated by finding them all under a single standardized location.

### Describe the current behavior

Arduino CLI places files in multiple folders under the temporary folder. For example:

```text
tmp/
├── arduino-core-cache/
├── arduino-sketch-33B7FB84F6F419C9D38C7A6D903E8E74/
├── arduino-sketch-8D027A8744130666283210C8D45F468C/
├── arduino-sketch-BE89418D95A99EBE12879C9FE4080CB4/
...
```

Since a separate folder is created for each sketch, this might be a significant number of folders.

🙁 It is more difficult for the user to configure exceptions in antivirus programs to not interfere with the compilation process.
🙁 It is more difficult for the user to do periodic manual clean ups of the temporary folder.

### Arduino CLI version

38a0dfd5eb9f08f8ff911e283fe2f643c20026ef

### Operating system

All

### Operating system version

Any

### Additional context

Originally requested by @sterretjeToo at https://github.com/arduino/arduino-ide/issues/1779

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-cli/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://arduino.github.io/arduino-cli/dev/installation/#nightly-builds)
- [X] My request contains all necessary details

### PR DESCRIPTION
**Pull request by giulianopz**
Fix #2028

Any temporary file will be persisted in a new dedicated folder named as "arduino" located in the OS-specific temporary directory, e.g.: 
```
/tmp/
├── arduino
│   ├── arduino-core-cache
│   │   └── core_arduino_avr_uno_640aa5b4d646262327bbea65c3462b39.a 
│   └── arduino-sketch-89E1544CF5F196B29E530BA6940B661F
│       ├── 06.ino.elf
│       ├── 06.ino.with_bootloader.bin
│       ├── build.options.json
│       ├── compile_commands.json
│       ├── core
│       │   ├── abi.cpp.d
[...]
```

This change is meant to make easier maintenance operations for both end-users and operating-systems.

## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [X] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [X] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)

## What kind of change does this PR introduce?

All code lines which reference the default tmp dir to store temporary files were changed to prefix "arduino-core-cache" and "arduino-sketch-[MD5]" dirs with "arduino/" as their parent.

## What is the current behavior?

See issue #2028.

## What is the new behavior?

All the temporary files will be put inside a new "arduino" folder in the default temporary directory of the OS.

## Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?

Not sure if this is a braking change with respect to older versions of the cli.

### CODE REVIEW

**Comment by pzygielo** [comment]
Will it work in multi-user environment for the second and next users, once the first temporary `arduino` directory is created?

**Comment by giulianopz** [comment]
> Will it work in multi-user environment for the second and next users, once the first temporary `arduino` directory is created?

The second user (and next ones) will not be allowed to compile the same sketch the first user built. But only if they can access the exact same sketch folder: the name of the tmp dir of a sketch ends with an MD5 hash value generated from the path of such dir.

This happens also with the current version of the cli, anyway.  

This PR just deals with grouping the two main dirs containing tmp files as requested in the referenced issue. You should open a new issue, I guess, if you think that this kind of collisions can really happen (I don't know, to be honest).

**Comment by pzygielo** [comment]
> This happens also with the current version of the cli (doesn't it?), anyway.

Now that you shown it I can see that `arduino-core-cache` locks other users already indeed.

File: arduino/sketch/sketch.go [hunk 1]
```diff
@@ -302,5 +302,5 @@ func GenBuildPath(sketchPath *paths.Path) *paths.Path {
 	}
 	md5SumBytes := md5.Sum([]byte(path))
 	md5Sum := strings.ToUpper(hex.EncodeToString(md5SumBytes[:]))
-	return paths.TempDir().Join("arduino-sketch-" + md5Sum)
+	return paths.TempDir().Join("arduino/arduino-sketch-" + md5Sum)
```
Comment (Bikappa) [inline arduino/sketch/sketch.go:305]: To allow cross platform compatibility in regards of the path separation, please use
```suggestion
	return paths.TempDir().Join("arduino", "sketch-" + md5Sum)
```
Also let's simplify the path a bit, by removing the double "arduino"

**Review by Bikappa — CHANGES_REQUESTED** [review]

**Review by Bikappa — DISMISSED** [review]

**Review by Bikappa — APPROVED** [review]

**Pull request merged**